nemo/collections/nlp/modules/common/megatron/retrieval_service.py

```diff
@@ -97,3 +207,272 @@ def get_knn(self, query: Union[List[str], str, torch.Tensor]):
+            end = time.time()
+            logging.info('convert Faiss db to GPU takes', end - beg)
+        self.index.nprobe = nprobe
+        self.tokenizer = tokenizer
+        self.ds = MMapRetrievalIndexedDataset(retrieval_index)
+        api = Api(self.app)
+        api.add_resource(
+            FaissRetrievalResource, '/knn', resource_class_args=[self.index, self.tokenizer, self.ds,],
+        )
+
+    def run(self, url, port=PORT_NUM):
+        threading.Thread(target=lambda: self.app.run(host=url, threaded=True, port=port)).start()
+
+
+class DynamicRetrievalResource(FaissRetrievalResource):
+    def __init__(
+        self, index, tokenizer, chunk_size, stride, store,
+    ):
+        self.index = index
+        self.tokenizer = tokenizer
+        self.chunk_size = chunk_size
+        self.stride = stride
+        self.pad_id = self.tokenizer.pad_id
+        self.ds = store
+
+    def put(self):
+        data = request.get_json()
+        if 'neighbors' in data:
+            sentences = data['sentences']
+            # do knn query
+            num_neighbors = data['neighbors']
+            with lock:  # Need to get lock to keep multiple threads from hitting code
+                neighbors = self.get_knn(sentences, num_neighbors)
+            return jsonify(neighbors.tolist())
+        elif 'reset' in data:
+            with lock:  # Need to get lock to keep multiple threads from hitting code
+                self.reset()
+            return "success"
+        else:
+            sentences = data['sentences']
+            add_eos = data['add_eos']
+            # update the index
+            with lock:  # Need to get lock to keep multiple threads from hitting code
+                self.add_docs_to_index(sentences, add_eos)
+            return "success"
+
+    def reset(self):
+        self.index.reset()
+        self.ds.reset()
+
+    def add_docs_to_index(self, docs: List[str], add_eos: bool = True):
+        """
+        Add documents to the Faiss index
+        Args:
+            docs: List[str], list of documents that is going to be added to the index
+            add_eos: bool, whether add the eos in the end
+        """
+        for doc in docs:
+            token_ids = self.tokenizer.text_to_ids(doc)
+            # append eos in the end
+            if add_eos:
+                token_ids.append(self.tokenizer.eos_id)
+            np_array = np.array(token_ids, dtype=np.int32)
+            padded_size = self.chunk_size - (len(np_array) % self.chunk_size)
+            # for retrieval database, added one more chunk in the end as padding
+            padded_size += self.chunk_size
+            np_array = np.pad(np_array, (0, padded_size), 'constant', constant_values=self.pad_id)
+            chunk_texts = []
+            for i in range(0, len(np_array), self.stride):
+                if i + 2 * self.chunk_size <= len(np_array):
+                    chunk = np_array[i : i + 2 * self.chunk_size]
```

Comment:
why is a chunk of size `2 * self.chunk_size`? A chunk is `64` tokens only right?

Reply:
Query chunk has 64 tokens. this is KNN context chunk, which has 128 tokens.

Reply:
Ah. I see.